includes/validation/class-amp-invalid-url-post-type.php

```diff
@@ -0,0 +1,1219 @@
+	 * @global WP $wp
+	 */
+	public static function store_validation_errors( $validation_errors, $url ) {
+		$post_slug = md5( $url );
+		$post      = get_page_by_path( $post_slug, OBJECT, self::POST_TYPE_SLUG );
+		if ( ! $post ) {
+			$post = get_page_by_path( $post_slug . '__trashed', OBJECT, self::POST_TYPE_SLUG );
+		}
+
+		// Since there are no validation errors and there is an existing $existing_post_id, just delete the post.
+		if ( empty( $validation_errors ) ) {
+			if ( $post ) {
+				wp_delete_post( $post->ID, true );
+			}
+			return null;
+		}
+
+		/*
+		 * The details for individual validation errors is stored in the amp_validation_error taxonomy terms.
+		 * The post content just contains the slugs for these terms and the sources for the given instance of
+		 * the validation error.
+		 */
+		$stored_validation_errors = array();
+
+		$terms = array();
+		foreach ( $validation_errors as $data ) {
+			/*
+			 * Exclude sources from data since not available unless sources are being obtained,
+			 * and thus not able to be matched when hashed.
+			 */
+			$sources = null;
+			if ( isset( $data['sources'] ) ) {
+				$sources = $data['sources'];
+			}
+
+			$term_data = AMP_Validation_Error_Taxonomy::prepare_validation_error_taxonomy_term( $data );
+			$term_slug = $term_data['slug'];
+			if ( ! isset( $terms[ $term_slug ] ) ) {
+
+				// Not using WP_Term_Query since more likely individual terms are cached and wp_insert_term() will itself look at this cache anyway.
+				$term = get_term_by( 'slug', $term_slug, AMP_Validation_Error_Taxonomy::TAXONOMY_SLUG );
+				if ( ! ( $term instanceof WP_Term ) ) {
+					$has_pre_term_description_filter = has_filter( 'pre_term_description', 'wp_filter_kses' );
+					if ( false !== $has_pre_term_description_filter ) {
+						remove_filter( 'pre_term_description', 'wp_filter_kses', $has_pre_term_description_filter );
+					}
+					$r = wp_insert_term( $term_slug, AMP_Validation_Error_Taxonomy::TAXONOMY_SLUG, wp_slash( $term_data ) );
+					if ( false !== $has_pre_term_description_filter ) {
+						add_filter( 'pre_term_description', 'wp_filter_kses', $has_pre_term_description_filter );
+					}
+					if ( is_wp_error( $r ) ) {
+						continue;
+					}
+					$term_id = $r['term_id'];
+					update_term_meta( $term_id, 'created_date_gmt', current_time( 'mysql', true ) );
+					$term = get_term( $term_id );
+				}
+				$terms[ $term_slug ] = $term;
+			}
+
+			$stored_validation_errors[] = compact( 'term_slug', 'sources' );
+		}
+
+		$post_content = wp_json_encode( $stored_validation_errors );
+		$placeholder  = 'amp_invalid_url_content_placeholder' . wp_rand();
+
+		// Guard against Kses from corrupting content by adding post_content after content_save_pre filter applies.
+		$insert_post_content = function( $post_data ) use ( $placeholder, $post_content ) {
+			$should_supply_post_content = (
+				isset( $post_data['post_content'] )
```

Comment:
Maybe this line could also check `$post_data['post_type']`, instead of the `isset( $post_data['post_type'] )` check [4 lines below](https://github.com/Automattic/amp-wp/pull/1093/files#diff-57518db0811ee918fee17d27459ed4ecR368):
```php
isset( $post_data['post_content'], $post_data['post_type'] )
```